src/utils/hook.ts

```diff
@@ -580,6 +580,69 @@ export function useParsedDate(timestamp: number): string {
   return parseDate(timestamp, now)
 }
 
+export const useCountdown = (targetDate: Date) => {
+  const countdownDate = new Date(targetDate).getTime()
+
+  const [countdown, setCountdown] = useState(countdownDate - new Date().getTime())
+
+  useEffect(() => {
+    const interval = setInterval(() => {
+      setCountdown(countdownDate - new Date().getTime())
+    }, 1000)
+
+    return () => clearInterval(interval)
+  }, [countdownDate])
+
+  const days = Math.floor(countdown / (1000 * 60 * 60 * 24))
+  const hours = Math.floor((countdown % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60))
+  const minutes = Math.floor((countdown % (1000 * 60 * 60)) / (1000 * 60))
+  const seconds = Math.floor((countdown % (1000 * 60)) / 1000)
+
+  return [days, hours, minutes, seconds]
+}
+
+export const useSingleHalving = (_halvingCount = 1) => {
+  const halvingCount = Math.max(Math.floor(_halvingCount) || 1, 1) // halvingCount should be a positive integer greater than 1.
+  const statistics = useStatistics()
+  const celebrationSkipKey = `having-celebration-${halvingCount}`
+  const celebrationSkipped = fetchCachedData(celebrationSkipKey) !== null
+  function skipCelebration() {
+    storeCachedData(celebrationSkipKey, true)
+  }
+
+  const currentEpoch = Number(statistics.epochInfo.epochNumber)
+  const targetEpoch = EPOCHS_PER_HALVING * halvingCount
+  const currentEpochUsedTime =
+    (Number(statistics.epochInfo.index) / Number(statistics.epochInfo.epochLength)) * THEORETICAL_EPOCH_TIME
+
+  const estimatedTime = (targetEpoch - currentEpoch) * THEORETICAL_EPOCH_TIME - currentEpochUsedTime
+  const estimatedDate = useMemo(() => new Date(new Date().getTime() + estimatedTime), [estimatedTime])
+
+  const haveDone = currentEpoch >= targetEpoch || new Date().getTime() > estimatedDate.getTime()
```

Comment:
I think `haveDone` should strictly base on `epoch` becuase it's time on the chain, `estimated time` is off-chain data and can be modified locally

Reply:
The purpose of this `havedone` is to show the celebration page, if it is only based on the `epoch` then it may produce a situation like that is, the theoretical time has gone, but the `epoch` is not 8760 (it may be a difference of only a few seconds or a few minutes) but then the front-end user will be shown that there are 0 seconds left, maybe we need to add an intermediate state for the countdown, such as `having coming soon` and so on, do you think that's OK?

Reply:
> The purpose of this `havedone` is to show the celebration page, if it is only based on the `epoch` then it may produce a situation like that is, the theoretical time has gone, but the `epoch` is not 8760 (it may be a difference of only a few seconds or a few minutes) but then the front-end user will be shown that there are 0 seconds left, maybe we need to add an intermediate state for the countdown, such as `having coming soon` and so on, do you think that's OK?

I think `halving coming soon` is better, it's precise and not confusing. The text could be `halving comes in seconds` so users will be informed it's really close, not minutes or hours.